src/core/io/platform/f4xx/ADCf4xx.cpp

```diff
@@ -84,109 +127,222 @@ void ADCf4xx::initADC(uint8_t num_channels) {
+        channel = CHANNEL_SET(1, 1, 0, ADC_CHANNEL_7);
         break;
     case Pin::PB_0:
-        adcChannel.Channel = ADC_CHANNEL_8;
+        channel = CHANNEL_SET(1, 1, 0, ADC_CHANNEL_8);
         break;
     case Pin::PB_1:
-        adcChannel.Channel = ADC_CHANNEL_9;
+        channel = CHANNEL_SET(1, 1, 0, ADC_CHANNEL_9);
         break;
     case Pin::PC_0:
-        adcChannel.Channel = ADC_CHANNEL_10;
+        channel = CHANNEL_SET(1, 1, 1, ADC_CHANNEL_10);
         break;
     case Pin::PC_1:
-        adcChannel.Channel = ADC_CHANNEL_11;
+        channel = CHANNEL_SET(1, 1, 1, ADC_CHANNEL_11);
         break;
     case Pin::PC_2:
-        adcChannel.Channel = ADC_CHANNEL_12;
+        channel = CHANNEL_SET(1, 1, 1, ADC_CHANNEL_12);
         break;
     case Pin::PC_3:
-        adcChannel.Channel = ADC_CHANNEL_13;
+        channel = CHANNEL_SET(1, 1, 1, ADC_CHANNEL_13);
         break;
     case Pin::PC_4:
-        adcChannel.Channel = ADC_CHANNEL_14;
+        channel = CHANNEL_SET(1, 1, 0, ADC_CHANNEL_14);
         break;
     case Pin::PC_5:
-        adcChannel.Channel = ADC_CHANNEL_15;
+        channel = CHANNEL_SET(1, 1, 0, ADC_CHANNEL_15);
         break;
     default:
+        channel = 0;
         break; // Should never get here
     }
 
+    // This checks if the pin being used supports the ADC being used
+    if (checkSupport(adcPeriph, channel)) {
+        // Masks channel back to proper value (Zero's out ADC information bits)
+        adcChannel.Channel = channel & 0x1F;
+    } else {
+        log::LOGGER.log(log::Logger::LogLevel::ERROR, "ADC %d DOES NOT SUPPORT PIN 0x%x!!", (adcNum + 1), pin);
+        // Causes HARD FAULT if pin does not support the ADC peripheral being used. THIS IS INTENTIONAL!
+        *((volatile int*) 0xFFFFFFFF) = 0; // This address is invalid
+    }
+
     // Subtract 1 because rank starts at 1
-    channels[rank - 1] = pin;
+    adcState->channels[rank - 1] = pin;
 
     adcChannel.Rank         = rank;
-    adcChannel.SamplingTime = ADC_SAMPLETIME_480CYCLES;
+    adcChannel.SamplingTime = ADC_SAMPLETIME_3CYCLES;
     adcChannel.Offset       = 0;
-    adcChannel.Offset       = 0x000;
 
-    HAL_ADC_ConfigChannel(&halADC, &adcChannel);
+    HAL_ADC_ConfigChannel(&adcState->halADC, &adcChannel);
+}
+
+inline bool ADCf4xx::checkSupport(ADCPeriph periph, uint32_t channel) {
+    // Checks if the channel contains the bit signifying the proper ADC peripheral support
+    switch (periph) {
+    case ADCPeriph::ONE:
+        return channel & (1 << ADC1SHIFT);
+    case ADCPeriph::TWO:
+        return channel & (1 << ADC2SHIFT);
+    case ADCPeriph::THREE:
+        return channel & (1 << ADC3SHIFT);
+    }
+}
+
+inline uint8_t ADCf4xx::getADCNum() {
+    switch (adcPeriph) {
+    case ADCPeriph::ONE:
+        return ADC1_SLOT;
+    case ADCPeriph::TWO:
+        return ADC2_SLOT;
+    case ADCPeriph::THREE:
+        return ADC3_SLOT;
+    }
+}
+
+void ADCf4xx::InitTimer() {
```

Comment:
Can you explain what exactly the timer is doing here?

Reply:
like, are you personally confused about it, or do you literally mean add a comment. im gonna assume its add a comment, but i can answer if u need

Reply:
I am personally confused, but a comment would probably help too. Put the comment in the header file though

Reply:
The timer limits the rate a which the ADC does a conversion, as without the timer slowing it down, it will do it so fast (even on slowest sample speed) that DMA interrupts will be called constantly (this is why the DMA ADC wouldn't work originally on f4). 
So this method initializes the timer that sends a Trigger Event to the ADC's to do a conversion. All ADC's are on the same timer and read at the same time.

Reply:
In the initADC() method on line 139, the ADC's external conversion trigger is set to Timer 8's trigger output. That, along with line 136 and 143 that sets the ADC to only do 1 conversion each trigger, means that it does exactly 1 conversion each timer cycle.

The ADC's listen for the timer trigger outputs. The timer has no awareness of the adc's. It just sends out a Trigger Output every cycle, unaware that anything is using it.

Reply:
You should add a nice big block comment around there explaining exactly that. My eyes usually glaze over when looking at the configuration macros, so I imagine a lot of people are going to miss that